### PR TITLE
Add wavelength input to LoadDNSLegacy

### LINKED ISSUE #24770: LoadDNSLegacy should get wavelength as an input
Presently, `LoadDNSLegacy` reads neutron wavelength and incident energy from the data file. However, data files often contain wrong wavelength and Ei. Thus, an option is needed to specify the correct wavelength. Ei can be than calculated and saved as a sample log.

### PR DESCRIPTION
**Pull request by mganeva**
**Description of work.**

This PR adds `Wavelength` input to `LoadDNSLegacy` algorithm. This is required for data reduction of measurements during the commissioning period until data acquisition software is fixed. 

**To test:**

Check that nothing has got broken. Simple script for testing:

**Diffraction mode data**

```python
# data file
datafile = 'dn134011vana.d_dat'

# Load dataset
ws = LoadDNSLegacy(datafile, Normalization='monitor', Wavelength=5.7)

# Check that sample logs are set correctly
run = ws.getRun()
print("Wavelength = {} Angstrom".format(run.getProperty('wavelength').value))  # should be 5.7
print("Ei = {} meV".format(round(run.getProperty('Ei').value, 2)))   # should be 2.52

# Check that X values are correct
import numpy as np
reference = np.zeros((24, 2))
reference.fill(5.7)
print("Are the X values correct? {}".format(np.allclose(ws.extractX(), reference, atol=1e-4)))
```
**TOF data**
```python
datafile = "dnstof.d_dat"
ws = LoadDNSLegacy(datafile, Normalization='no', Wavelength=5.7)

# check sample logs
run = ws.getRun()
print("Wavelength = {} Angstrom".format(run.getProperty('wavelength').value))  # should be 5.7
print("Ei = {} meV".format(round(run.getProperty('Ei').value, 2)))   # should be 2.52
print("TOF1 = {} microsecond".format(round(run.getProperty('TOF1').value, 2)))   # should be 576.33

# check whether TOF at zero time channel is equal to TOF1
import numpy as np
print(np.round(ws.extractX()[:,0], 2) == 576.33)
```

Feel free to modify wavelength and see whether Ei and TOF1 changes respectively. For DNS source-sample distance L1=0.4 m. Sample-detector distance L2=0.8 m.

Fixes #24770 . 

*This does not require release notes because it is a minor change*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/LoadDNSLegacy.py [hunk 1]
```diff
@@ -132,6 +136,13 @@ def PyExec(self):
             self.log().error(message)
             raise RuntimeError(message)
 
+        if wavelength < 0.01:
+            wavelength = metadata.wavelength
+
+        # calculate incident energy, since given in the data file is wrong
+        velocity = h/(m_n*wavelength*1e-10)   # m/s
+        incident_energy = 0.5*m_n*velocity*velocity/1.602176487e-22  # meV
```
Comment (soininen) [inline Framework/PythonInterface/plugins/algorithms/LoadDNSLegacy.py]: One could use `scipy.constants` and `velocity**2` here.

File: Framework/PythonInterface/plugins/algorithms/LoadDNSLegacy.py [hunk 2]
```diff
@@ -64,6 +64,9 @@ def PyInit(self):
 
         self.declareProperty(name="ElasticChannel",defaultValue=0,validator=IntBoundedValidator(lower=0),
                              doc="Time channel number where elastic peak is observed. Only for TOF data.")
+
+        self.declareProperty(name="Wavelength",defaultValue=0.0,validator=FloatBoundedValidator(lower=0.0),
```
Comment (soininen) [inline Framework/PythonInterface/plugins/algorithms/LoadDNSLegacy.py]: It is possible to use `EMPTY_DBL` as an indication of 'no value given'. I don't know which one would be nicer for the users, though.

**Review by soininen — CHANGES_REQUESTED** [review]
Just a few quick comments on the code before I test this.

File: Framework/PythonInterface/plugins/algorithms/LoadDNSLegacy.py (hunk 2 above)
Comment (mganeva) [reply Framework/PythonInterface/plugins/algorithms/LoadDNSLegacy.py]: I think, zero should be easier to understand for users. Anyway, it is nonphysical value for the wavelength.

File: Framework/PythonInterface/plugins/algorithms/LoadDNSLegacy.py (hunk 1 above)
Comment (mganeva) [reply Framework/PythonInterface/plugins/algorithms/LoadDNSLegacy.py]: x*x is about 1.5 times faster, than x**2. It is not critical for this algorithm, but I'd prefer leave velocity*velocity.

**Comment by soininen** [comment]
@mganeva Sorry for not mentioning this earlier, but I think this still needs release notes as it changes the user-facing API by adding a new property to the algorithm. I do not think such changes can be considered 'minor'.

**Review by soininen — APPROVED** [review]

**Comment by soininen** [comment]
Tested this as instructed and the new Wavelength property seems to work as promised. The code changes look OK, test have been added and release notes updated. Everything looks good to me.

**Pull request merged**